include/pybind11/stl.h

```diff
@@ -305,6 +314,59 @@ template<> struct type_caster<std::experimental::nullopt_t>
     : public void_caster<std::experimental::nullopt_t> {};
 #endif
 
+#if defined(PYBIND11_HAS_FILESYSTEM)
+template<> struct type_caster<std::filesystem::path> {
+
+    static PyObject* unicode_from_fs_native(const std::string& w) {
+        return PyUnicode_DecodeFSDefaultAndSize(w.c_str(), ssize_t(w.size()));
+    }
+
+    static PyObject* unicode_from_fs_native(const std::wstring& w) {
+        return PyUnicode_FromWideChar(w.c_str(), ssize_t(w.size()));
+    }
```

Comment:
We could merge these two. `if constexpr` would be fine, because this caster already depends on C++17. My first attempt was:

```c++
static PyObject* unicode_from_fs_native(const std::filesystem::path::string_type& w) {
    if constexpr (std::is_same_v<std::filesystem::path::string_type, std::string>) {
        return PyUnicode_DecodeFSDefaultAndSize(w.c_str(), ssize_t(w.size()));
    } else if (!std::is_same_v<std::filesystem::path::string_type, std::wstring>) {
        return PyUnicode_FromWideChar(w.c_str(), ssize_t(w.size()));
    }
}
```

That doesn't work, because the compiler can still see that `w.c_str()` doesn't match both `char*` and `wchar_t*`. So we have to make it a template...

```c+
template<typename T = std::filesystem::path::string_type>
PyObject* unicode_from_fs_native(const T& w) {
    if constexpr (std::is_same_v<T, std::string>) {
        return PyUnicode_DecodeFSDefaultAndSize(w.c_str(), ssize_t(w.size()));
    } else if (std::is_same_v<T, std::wstring>) {
        return PyUnicode_FromWideChar(w.c_str(), ssize_t(w.size()));
    }
}
```

The problem with the above is that we really don't want `T` to be deduced - we want the default to always apply. The language calls this a [non-deduced context](https://eel.is/c++draft/temp.deduct.type#5) and `std::type_identity_t<T>` is exactly a non-deduced `T`.

```c++
template<typename T = std::filesystem::path::string_type>
PyObject* unicode_from_fs_native(const std::type_identity_t<T>& w) {
    if constexpr (std::is_same_v<T, std::string>) {
        return PyUnicode_DecodeFSDefaultAndSize(w.c_str(), ssize_t(w.size()));
    } else if (std::is_same_v<T, std::wstring>) {
        return PyUnicode_FromWideChar(w.c_str(), ssize_t(w.size()));
    }
}
```

This version finally works, but `std::type_identity<T>` is C++20... Final result, with type identity reimplemented, can be found here: https://godbolt.org/z/hPYdq7

Reply:
I basically went through the same motions until realizing the need for a non-deduced context, and not being aware of `type_identity` (thanks for pointing it out to me) I decided a good old overload was good enough.  I'd say in this case the overload is actually more legible than the `if constexpr` + backport `type_identity` version?
Still, if you really prefer that, let me know and I'll make the changes.

Reply:
Yeah, I'm also not convinced that template + `if constexpr` + backport of `type_identity` is worth it. Keep it as is.

Reply:
Actually I found a nicer way to achieve this, using a lambda taking auto as parameter (which effectively gives us something like a non-deduced context).

Reply:
Can I be annoying and say that this lambda too me a lot longer to parse than selecting the right overload?

Reply:
I don't really care about which approach to use, but I'll let you and @bstaletic decide together which one you want once and for all instead of switching back and forth ;)

Reply:
Sure. Up to @bstaletic. I am/was not saying it should be changed, but just surprised that this is considered cleaner than the overload :-)

Reply:
I'm fine either way. If @YannickJadoul finds the overload version easier to read, I guess we should do that.

@anntzer apologies for nagging.

Reply:
back to good old overloading.

Reply:
Sorry for nagging as well, and making you revert this, @anntzer!